addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/discovery/DigiplexDiscoveryService.java

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.digiplex.discovery;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryServiceCallback;
+import org.eclipse.smarthome.config.discovery.ExtendedDiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.digiplex.DigiplexBindingConstants;
+import org.openhab.binding.digiplex.communication.AreaLabelRequest;
+import org.openhab.binding.digiplex.communication.AreaLabelResponse;
+import org.openhab.binding.digiplex.communication.DigiplexMessageHandler;
+import org.openhab.binding.digiplex.communication.DigiplexRequest;
+import org.openhab.binding.digiplex.communication.ZoneLabelRequest;
+import org.openhab.binding.digiplex.communication.ZoneLabelResponse;
+import org.openhab.binding.digiplex.handler.DigiplexBridgeHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Service for discovering things on Digiplex alarm systems
+ *
+ * @author Robert Michalak - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class DigiplexDiscoveryService extends AbstractDiscoveryService
```

Comment:
Please make this a real OSGi service (`@Component`) with `immediate=false`.
There is a new pattern to be used for discovery services that are bound to handlers:
1. Return the class in your handlers `getServices()`.
2. Implement the required interface

Reply:
Do you have any example of such a usage? 

My understanding is that I should no longer extend `AbstractDiscoveryService` but implement `ThingHandlerService`, so that means I should no longer depend on `startScan` and `stopScan` methods from the former class but on `activate` and `deactivate` methods from the latter one?

Reply:
You extend abstract discovery service and you implement the mentioned interface.

Unfortunately there is no openHAB documentation written yet, it is only mentioned on the old eclipse smarthome developer section.

Have a look for that interface here in the repository via the GitHub search. I will also look for one, just not now.

Reply:
That was my first impression as well, but unfortunately I received this error when doing so:

`The inherited method AbstractDiscoveryService.deactivate() cannot hide the public abstract method in ThingHandlerService`

I was able to overcome this by overriding `decativate` method to artificially change its visibility to `public`, but it's not a very elegant long term solution (to require it for every discovery service bound to a Thing).

There is one more missing piece in this area: what to do with `DigiplexHandlerFactory#registerDigiplexDiscoveryService` method? Is it still needed? My understanding of framework changes is that discovery services are now registered/removed automatically(?)

Reply:
> I was able to overcome this by overriding decativate method to artificially change its visibility to public, but it's not a very elegant long term solution (to require it for every discovery service bound to a Thing).

Actually that is a requirement for all newly written services to expose activate/deactivate. The underlying/inherited ones are hidden from the OSGi framework. 

> Is it still needed? 

No. Can be removed. That is automatically handled.